verifier/src/lib.rs

```diff
@@ -17,11 +21,66 @@ pub use winterfell::{StarkProof, VerifierError};
 /// # Errors
 /// Returns an error if the provided proof does not prove a correct execution of the program.
 pub fn verify(
-    program_hash: [u8; 32],
-    public_inputs: &[u128],
-    outputs: &[u128],
+    program_hash: Digest,
+    stack_inputs: &[u64],
+    stack_outputs: &[u64],
     proof: StarkProof,
-) -> Result<(), VerifierError> {
-    let pub_inputs = PublicInputs::new(program_hash, public_inputs, outputs);
-    winterfell::verify::<ProcessorAir>(proof, pub_inputs)
+) -> Result<(), VerificationError> {
+    if stack_inputs.len() > MIN_STACK_DEPTH {
+        return Err(VerificationError::TooManyInputValues(
+            MIN_STACK_DEPTH,
+            stack_inputs.len(),
+        ));
+    }
+
+    // convert stack inputs to field elements
+    let mut stack_input_felts = Vec::with_capacity(stack_inputs.len());
+    for &input in stack_inputs.iter().rev() {
+        stack_input_felts.push(
+            input
+                .try_into()
+                .map_err(|_| VerificationError::InputNotFieldElement(input))?,
+        );
+    }
+
+    if stack_outputs.len() > MIN_STACK_DEPTH {
+        return Err(VerificationError::TooManyOutputValues(
+            MIN_STACK_DEPTH,
+            stack_outputs.len(),
+        ));
+    }
+
+    // convert stack outputs to filed elements
+    let mut stack_output_felts = Vec::with_capacity(stack_outputs.len());
+    for &output in stack_outputs.iter().rev() {
+        stack_output_felts.push(
+            output
+                .try_into()
+                .map_err(|_| VerificationError::InputNotFieldElement(output))?,
+        );
+    }
```

Comment:
I thought about this but the fact that we return a different error makes it less attractive to do that.

But - your comment made me notice that I'm returning a wrong error here: should have been `OutputNotFieldElement` rather than `InputNotFieldElement`.

Reply:
Yep, if the errors are different this makes sense